README.md
# gpt-summarizer
### Extract text, summarize each section w/ GPT, and provide a summarized outline of a paper/article

This script extracts text from a given file or URL and splits it into sections. It then uses OpenAI's tokenizer to encode the text as a sequence of tokens. It writes the extracted text to an output file and writes each section to a separate text file. It also generates a summary for each subsection and writes the summary to a summary file. If there are multiple summary files for a section, it generates a combined section summary.

The script expects a PDF or HTML file path, or an HTML URL, to be passed as a command line argument. It extracts the text from the PDF file, splits the text into sections, and uses the tiktoken.get_encoding function to encode the text as a sequence of tokens using the "gpt2" encoding. It writes the extracted text to an output file using the base_name of the file and the .txt extension.

It processes each section by first using the split_section_into_subsections function to split the section into subsections based on HTML section headings or numbered section headings. If necessary, it further splits any subsections into paragraphs and recombine adjacent ones until they exceed 1000 tokens. It processes each resulting section/part with InstructGPT (text-davinci-003) to generate a summary, and writes each section summary to a summary file. If there are multiple summary files for a section, it generates a combined section summary by concatenating the summaries of the individual subsections.

It then performs one final round of summarization across all the lower-level summaries, to produce an overall summary of the paper/article.

The intended use is that both the overall summary and the subsection summaries (in order) are worth reading to determine whether to spend the time reading the entire article/paper, or specific sections of it.

You can also specify an optional second positional argument to have the summaries generated in the specified language: to do so it adds `" Please use "+sys.argv[2]+" language for the output."` to the prompt if the argument is present.

## Usage

tl;dr:
```
brew install libmagic 
pip -r requirements.txt
export OPENAI_API_KEY=<your OpenAI API key>
python summarize.py paper.pdf
python summarize.py https://path/to/article
python summarize.py https://path/to/article Spanish
```

### Usage details:

Requires an OpenAI API key:
 - If you haven't already done so to get access to ChatGPT, sign up for an account at OpenAI.com
 - Go to https://openai.com/api/ and Log in
 - Go to https://beta.openai.com/account/api-keys
 - Create a new secret key
 
This OPENAI_API_KEY should be set as an environment variable:
`export OPENAI_API_KEY=<your OpenAI API key>`

For now, you also have to manually pip install pdfminer, html2text, tiktoken, and openai. ~~(I'd welcome a PR to get this repo set up to use setup requirements to support `pip install -e .`)~~


## Examples

https://github.com/scottleibrand/gpt-summarizer/tree/main/examples

### Summary of OurWorldInData AI Impact article  
https://ourworldindata.org/ai-impact

```
~/src/gpt-summarizer $ ./summarize.py https://ourworldindata.org/ai-impact
HTML path: /tmp/ai-impact.html
URL: https://ourworldindata.org/ai-impact
/tmp/ai-impact.html
/tmp/ai-impact
Text extracted from https://ourworldindata.org/ai-impact and written to /tmp/ai-impact.full.txt
Total token count: 8038
Header:  Title-Abstract
Title-Abstract (10563 characters, 2493 tokens) written to /tmp/ai-impact.TitleAbstract.full.txt
Summary written to /tmp/ai-impact.TitleAbstract.summary.txt
No summary files found for section TitleAbstract
Header:  ##### A future of human-level or transformative AI?
##### A future of human-level or transformative AI? (1140 characters, 258 tokens) written to /tmp/ai-impact.AfutureofhumanlevelortransformativeAI.full.txt
Summary written to /tmp/ai-impact.AfutureofhumanlevelortransformativeAI.summary.txt
No summary files found for section AfutureofhumanlevelortransformativeAI
Header:  #### What is at stake as artificial intelligence becomes more powerful?
#### What is at stake as artificial intelligence becomes more powerful? (4677 characters, 1059 tokens) written to /tmp/ai-impact.Whatisatstakeasartificialintelligencebecomesmorepowerful.full.txt
Summary written to /tmp/ai-impact.Whatisatstakeasartificialintelligencebecomesmorepowerful.summary.txt
No summary files found for section Whatisatstakeasartificialintelligencebecomesmorepowerful
Header:  #### How can we make sure that the development of AI goes well?
#### How can we make sure that the development of AI goes well? (3427 characters, 792 tokens) written to /tmp/ai-impact.HowcanwemakesurethatthedevelopmentofAIgoeswell.full.txt
Summary written to /tmp/ai-impact.HowcanwemakesurethatthedevelopmentofAIgoeswell.summary.txt
No summary files found for section HowcanwemakesurethatthedevelopmentofAIgoeswell
Header:  ### Endnotes
### Endnotes (10116 characters, 2573 tokens) written to /tmp/ai-impact.Endnotes.full.txt
Summary written to /tmp/ai-impact.Endnotes.summary.txt
No summary files found for section Endnotes
Header:  ### Reuse this work freely
### Reuse this work freely (2764 characters, 853 tokens) written to /tmp/ai-impact.Reusethisworkfreely.full.txt
Summary written to /tmp/ai-impact.Reusethisworkfreely.summary.txt
No summary files found for section Reusethisworkfreely
No abstract found for /tmp/ai-impact
Concatenated 0 summaries into a single summary with 2 characters and 1 tokens
Concatenated subsection summaries have less than 500 tokens, reading in all summaries
Overall summary written to /tmp/ai-impact.overall_summary.txt
~/src/gpt-summarizer $
```

https://github.com/scottleibrand/gpt-summarizer/blob/main/examples/ai-impact.overall_summary.txt

#### OWID Overall Summary

This article discusses the potential implications of artificial intelligence (AI) becoming a reality. It explains why it is difficult to take the prospect of a world transformed by AI seriously, and how to develop an idea of what the future of AI might look like. It compares the potential of transformative AI to the agricultural and industrial revolutions, and suggests that it could represent the introduction of a similarly significant general-purpose technology. The article also looks at the advantages and disadvantages of comparing machine and human intelligence, and introduces the concept of transformative AI, which is defined by the impact this technology would have on the world. It is noted that transformative AI could be developed before human-level AI, and that the timeline for when either of these levels of AI might be achieved is difficult to predict.

The article also looks at the potential risks and benefits of AI becoming more powerful. It is clear that AI can already cause harm when used maliciously, such as in politically-motivated disinformation campaigns or to enable mass surveillance. AI can also cause unintended harm, such as when an AI system falsely accused 26,000 parents of making fraudulent claims for child care benefits in the Netherlands. As AI becomes more powerful, the potential negative impacts could become much larger, such as mass labor displacement, extreme concentrations of power and wealth, and totalitarianism. Additionally, there is the risk of an AI system escaping human control and harming humans, known as the alignment problem. This risk is difficult to foresee and prevent, and could lead to an extreme catastrophe. On the other hand, AI could lead to positive developments such as cleaner energy, the replacement of unpleasant work, and better healthcare. The stakes are high with this technology, and reducing the negative risks and solving the alignment problem could mean the difference between a healthy, flourishing, and wealthy future for humanity – and the destruction of the same.

The article also looks at the difference between human-level AI and transformative AI, and the potential timeline for when either of these levels of AI might be achieved. It is noted that transformative AI could be developed before human-level AI, and that the timeline for when either of these levels of AI might be achieved is difficult to predict. Additionally, the article provides information about the licenses and permissions associated with Our World in Data's visualizations, data, code, and articles. Finally, the article looks at the concept of human-level AI, which is defined as a software system that can carry out at least 90% or 99% of all economically relevant tasks that humans carry out. It also looks at the closely related terms Artificial General Intelligence, High-Level Machine Intelligence, Strong AI, or Full AI, which are sometimes defined in similar, yet different ways. The section also looks at the difficulty of comparing machine and human intelligence, and the potential risks of AI systems, such as AI-enabled disinformation campaigns and mass surveillance by governments. It also looks at the incentives for developing powerful AI, and the potential for it to lead to positive developments. Finally, the section looks at the early warnings of Alan Turing and Norbert Wiener about the alignment problem, and Toby Ord's projection that AI could be developed by 2040.

#### OWID Section Summaries

`ls -rt *summary.txt | while read file; do echo -n $file; cat $file; echo; echo; done`

ai-impact.AfutureofhumanlevelortransformativeAI.summary.txt

This section discusses the difference between human-level AI and transformative AI, and the potential timeline for when either of these levels of AI might be achieved. It is noted that transformative AI could be developed before human-level AI, and that the timeline for when either of these levels of AI might be achieved is difficult to predict. The article provides a link to a companion article which gives an overview of what researchers in this field currently believe about the timeline for AI development.

ai-impact.Endnotes.summary.txt

This section discusses the concept of human-level AI, which is defined as a software system that can carry out at least 90% or 99% of all economically relevant tasks that humans carry out. It also looks at the closely related terms Artificial General Intelligence, High-Level Machine Intelligence, Strong AI, or Full AI, which are sometimes defined in similar, yet different ways. The section also looks at the difficulty of comparing machine and human intelligence, and the potential risks of AI systems, such as AI-enabled disinformation campaigns and mass surveillance by governments. It also looks at the incentives for developing powerful AI, and the potential for it to lead to positive developments. Finally, the section looks at the early warnings of Alan Turing and Norbert Wiener about the alignment problem, and Toby Ord's projection that AI could be developed by 2040.

ai-impact.HowcanwemakesurethatthedevelopmentofAIgoeswell.summary.txt

Making sure that the development of artificial intelligence (AI) goes well is a crucial question for humanity. Currently, resources dedicated to AI are mostly focused on speeding up its development, while efforts to increase its safety are under-resourced. This neglect of AI safety work means that individuals have a good chance to make a positive difference if they dedicate themselves to this problem. However, it needs more than individual efforts; society needs to become knowledgeable about the technology and understand what is at stake. Our World in Data is doing its part to enable a better informed public conversation on AI and the future we want to live in.

ai-impact.Reusethisworkfreely.summary.txt

This section provides information about the licenses and permissions associated with Our World in Data's visualizations, data, code, and articles. All of Our World in Data's work is open access under the Creative Commons BY license, and all software and code is open source under the MIT license. Data produced by third parties is subject to the license terms from the original third-party authors. Additionally, Our World in Data's charts can be embedded in any site, and the project is a part of the Global Change Data Lab, a registered charity in England and Wales. A full legal disclaimer is also provided.

ai-impact.TitleAbstract.summary.txt

This section discusses the potential implications of artificial intelligence (AI) becoming a reality. It explains why it is difficult to take the prospect of a world transformed by AI seriously, and how to develop an idea of what the future of AI might look like. It also looks at the advantages and disadvantages of comparing machine and human intelligence, and introduces the concept of transformative AI, which is defined by the impact this technology would have on the world. It compares the potential of transformative AI to the agricultural and industrial revolutions, and suggests that it could represent the introduction of a similarly significant general-purpose technology.

ai-impact.Whatisatstakeasartificialintelligencebecomesmorepowerful.summary.txt


CHANGELOG.md
# CHANGELOG

* _main() as an entry point:_ Most of the execution logic/sequencing is currently done outside of a function. You properly handle the condition in which the script is run from the command line by checking the invocation context (if __name__ == '__main__': ), but I've moved much of that sequencing to a discrete function (main()) which you then call from your conditional as it exists today. This allows you to consider using your application as a module and importing it from another script/application with only a few minor changes. But more importantly I've found it forces me to be sure my logic is sound and linear and allows for a third-party to more easily follow the story line of what I'm trying to do.
* _argparse instead of argv_: I moved your argument handling to the argparse library and you can see it fits naturally as the start of the program at our new entry point.  It should be self-explanatory if you run _python summarize.py -h_ but I wanted to be extra mindful to not change the functionality or your intentional choices for handling user input. You'll note this is the start of my mission against making calls directly to the OS and allowing libraries to do that work for us. :)
  * _-b (basename):_ I wasn't sure if you wanted to give the user the opportunity to set a base_name as an alternative to inferring it from the URL/filename.  So there's an optional parameter that will set the base_name to user input.  NB: I got lazy and passed the variable around a bit to avoid having to make two calls to helper functions - I guess the right way to do it would be to create an object of type Document and read its state and ignore its base_name if we set it manually, but alas, this is good enough for government work 
  
  * _mutual exclusion:_ It'd not be difficult to fire off a list of articles/documents you wanted to summarize but again, since I didn't want to change the original functionality, I allowed the user to specify -f or -u, but not both
* _requests vs. curl_:  I had a debate with myself as to whether or not to use the requests library
    or the system-call to curl as you've done below.  i ended up siding with requests because:
  * some OS-level issue with curl e.g. nonstandard implementation, not in path, etc. and 
  * requests gives us some clue into the type of document returned rather than having to rely on filename
* _Pathlib vs. os.path: For similar reasons to what I mentioned earlier, I like using Pathlib to handle navigating directory structures and so I used a pretty trivial example in your code. It's more powerful than I demonstrate in that you can test permissions, etc. before writing.  In your case I changed the destination of the downloaded file to the user's home directory since they may not have access to the rootlevel /tmp but feel free to change to whatever you'd like obviously.

_Miscellany_

* I used a bit of lazy hackery to get around a condition I found in which a URL ended with a trailing slash, nothing after it, yet supplied a PDF file.  I opted to rely more on the guesswork from libmagic and PDFReader vs. trying to intuit the type of file from the filename itself
* I tried to take about 300 lines of code in the latter half of your original execution logic and put them into separate functions as best I could ,but I did **not** continue adjusting the design pattern/syntax and rather just copied and pasted.
* I've mentioned this a bunch of times now but as I got near the end I couldn't help but hate the fact that I don't have more time because I ended up passing variables around like crazy whereas constructing an object of type Article or something where we can access its state from anywhere.  For another day :)

# BROKEN
I left the attribution portion broken because I lost track of the original URL variable :)  It's commented out.




extract_text.py
#!/usr/bin/python

"""
This script contains functions for extracting text from a PDF file, splitting
the text into sections, and splitting sections into subsections. The
extract_text_from_pdf() function takes in a file path to a PDF and returns a
string of the extracted text. The split_into_sections() function takes in a
string of text and uses a regular expression to split it into a list of tuples,
where each tuple contains a section header and the corresponding text. The
split_section_into_subsections() function takes in a section header, the
corresponding text, and an encoder object and splits the section into smaller
parts, each of which is returned as a tuple containing a subsection header and
the corresponding text. The split_subsection_into_paragraphs() function takes
in a subsection header, the corresponding text, the encoder object, and a
maximum number of tokens and splits the subsection into smaller parts, each of
which is returned as a tuple containing a paragraph header and the
corresponding text.

This script was written by ChatGPT with direction by Scott Leibrand,
then edited by Scott Leibrand w/ CoPilot and ChatGPT.
"""

import sys
import re
import os
import openai
import glob

import tiktoken
from io import StringIO
from pdfminer.pdfinterp import PDFResourceManager, PDFPageInterpreter
from pdfminer.converter import TextConverter
from pdfminer.layout import LAParams
from pdfminer.pdfpage import PDFPage

def extract_text_from_pdf(pdf_path):
    """Extracts the text from a PDF file and returns it as a string.

    Parameters:
    pdf_path (str): The file path to the PDF file.

    Returns:
    str: The extracted text.
    """
    with open(pdf_path, 'rb') as fh:
        # Create a PDF resource manager object that stores shared resources
        rsrcmgr = PDFResourceManager()

        # Create a StringIO object to store the extracted text
        output = StringIO()

        # Create a TextConverter object to convert PDF pages to text
        device = TextConverter(rsrcmgr, output, laparams=LAParams())

        # Create a PDF page interpreter object
        interpreter = PDFPageInterpreter(rsrcmgr, device)

        # Process each page contained in the PDF document
        for page in PDFPage.get_pages(fh, caching=True, check_extractable=True):
            interpreter.process_page(page)

        # Get the extracted text as a string and close the StringIO object
        text = output.getvalue()
        output.close()

        # Close the PDF file and text converter objects
        device.close()

    # Remove ^L page breaks from the text
    text = text.replace('\x0c', '\n')

    return text


def split_into_sections(text):
    """Splits a string of text into a list of tuples, where each tuple contains a section header and the corresponding text.

    Parameters:
    text (str): The input text to split into sections.

    Returns:
    list: A list of tuples, where each tuple contains a section header and the corresponding text.
    """
    # Use a regular expression to match the "References" section
    pattern = r'(References[^\n]*)\n'
    match = re.search(pattern, text)
    if match:
        # Remove the "References" section and everything that follows
        text = text[:match.start()]

    # Use a regular expression to split the text into sections
    #pattern = r'\n\n((?:\d+[\.:]|(?:Introduction|Methods|Materials and Methods|Results|Discussion|References))\s+[^\n]+)\n\n'
    pattern = r'\n\n(\d+[\.:]\s+[^\n]+)\n\n'
    sections = re.split(pattern, text)

    # Extract the section headers and their corresponding text
    headers = ["Title-Abstract"]
    content = []
    for i, section in enumerate(sections):
        if i % 2 == 0:
            # This is a section of content
            content.append(section)
        else:
            # This is a section header
            headers.append(section)

    # Zip the section headers and content together
    sections = list(zip(headers, content))

    #print(headers)

    return sections

def split_section_into_subsections(section_header, section_content, enc, max_tokens=3000):
    """Splits a section of text into smaller parts, each of which is returned
    as a tuple containing a subsection header and the corresponding text.

    Parameters:
    section_header (str): The header for the section to be split.
    section_content (str): The content of the section to be split.
    enc (object): An encoder object used to encode the section content as a sequence of tokens.
    max_tokens (int, optional): The maximum number of tokens allowed in each subsection. Default is 3000.

    Returns:
    list: A list of tuples, where each tuple contains a subsection header and the corresponding text.
    """
    # Encode the section content as a sequence of tokens
    tokens = enc.encode(section_content)

    if len(tokens) <= max_tokens:
        # The section does not need to be split into subsections
        return [(section_header, section_content)]

    # Split the section into numbered subsections
    pattern = r'\n\n(\d+\.\d+[\.:]\s+[^\n]+)\n\n'
    subsections = re.split(pattern, section_content)

    # Extract the subsection headers and their corresponding text
    headers = [f"{section_header.split('.')[0]}. Section intro"]
    content = []
    for i, subsection in enumerate(subsections):
        if i % 2 == 0:
            # This is a subsection of content
            content.append(subsection)
        else:
            # This is a subsection header
            headers.append(subsection)

    # Zip the subsection headers and content together
    subsections = list(zip(headers, content))


    # Split any subsections that are still too long into smaller parts
    result = []
    for header, content in subsections:
        parts = split_subsection_into_paragraphs(header, content, enc, max_tokens)
        result.extend(parts)

    return result

    #return subsections

def split_subsection_into_parts_broken(subsection_header, subsection_content, enc, max_tokens=3000):
    # Encode the subsection content as a sequence of tokens
    tokens = enc.encode(subsection_content)

    if len(tokens) <= max_tokens:
        # The subsection does not need to be split into parts
        return [(subsection_header, subsection_content)]

    # Split the subsection into parts
    start = 0
    parts = []
    while start < len(tokens):
        # Calculate the size of the next part
        end = start + max_tokens

        # Find the nearest newline boundary within the part
        print(enc)
        while end < len(tokens) and tokens[end] != enc.encoder['\n']:
            end += 1

        # Extract the part
        part_tokens = tokens[start:end]
        part_content = enc.decode(part_tokens)

        # Add the part to the list of parts
        parts.append((subsection_header, part_content))

        # Update the start index
        start = end + 1

    return parts

def split_subsection_into_paragraphs(subsection_header, subsection_content, enc, max_tokens=3000):
    # Encode the subsection content as a sequence of tokens
    tokens = enc.encode(subsection_content)

    if len(tokens) <= max_tokens:
        # The subsection does not need to be split into parts
        return [(subsection_header, subsection_content)]

    # Split the subsection into parts
    start = 0
    parts = []
    while start < len(tokens):
        # Calculate the size of the next part
        end = start + max_tokens

        # Find the nearest newline boundary within the part
        newline_pos = subsection_content[start:end].find('\n\n')
        if newline_pos != -1:
            end = start + newline_pos

        # Extract the part
        part_tokens = tokens[start:end]
        part_content = enc.decode(part_tokens)

        # Add the part to the list of parts
        parts.append((subsection_header, part_content))

        # Update the start index
        start = end + 2

    return parts

def combine_subsections(subsections):
    # Initialize the list of combined subsections
    combined_subsections = []

    # Initialize the current combined subsection
    current_subsection_header = ""
    current_subsection_content = ""
    current_subsection_tokens = 0

    # Iterate through the subsections
    for header, content in subsections:
        # Encode the content as a sequence of tokens
        tokens = enc.encode(content)

        # If the current combined subsection has less than 1000 tokens and the current subsection has less than 1000 tokens, combine them
        if current_subsection_tokens + len(tokens) < 2000 and len(tokens) < 1000:
            # Update the current combined subsection header
            if current_subsection_header == "":
                current_subsection_header = header
                current_subsection_content = header + "\n"
            else:
                if current_subsection_header != header:
                    current_subsection_content += "\n\n" + header + "\n"
                #current_subsection_header += header

            # Update the current combined subsection content
            current_subsection_content += content

            # Update the current combined subsection token count
            current_subsection_tokens += len(tokens)
        else:
            # Add the current combined subsection to the list of combined subsections
            combined_subsections.append((current_subsection_header, current_subsection_content))

            # Reset the current combined subsection
            current_subsection_header = header
            current_subsection_content = header + "\n" + content
            current_subsection_tokens = len(tokens)

    # Add the final combined subsection to the list of combined subsections
    combined_subsections.append((current_subsection_header, current_subsection_content))

    return combined_subsections

import openai
import os

def generate_summary(content, prompt):
    # Get the API key from the environment variable
    api_key = os.environ["OPENAI_API_KEY"]
    openai.api_key = api_key

    # Set the model to use
    model_engine = "text-davinci-003"

    # Set the temperature for sampling
    temperature = 0

    # Set the max token count for the summary
    max_tokens = 1000

    # Generate completions
    completions = openai.Completion.create(
        engine=model_engine,
        prompt=prompt,
        max_tokens=max_tokens,
        temperature=temperature
    )

    # Get the summary from the first completion
    summary = completions.choices[0].text

    return summary


if __name__ == '__main__':
    # Get the PDF file path from the command line arguments
    pdf_path = sys.argv[1]

    # Extract the text from the PDF file
    text = extract_text_from_pdf(pdf_path)

    # Split the text into sections
    sections = split_into_sections(text)

    # Use tiktoken to encode the text as a sequence of tokens
    enc = tiktoken.get_encoding("gpt2")
    tokens = enc.encode(text)

    # Get the base name of the output file
    base_name, _ = os.path.splitext(pdf_path)

    # Write the extracted text to the output file
    with open(base_name + ".txt", 'w') as f:
        f.write(text)

    print(f"Text extracted from {pdf_path} and written to {base_name}.txt")


    print(f"Total token count: {len(tokens)}")

    # Write each section to a separate text file
    for header, content in sections:
        print("Header: ", header)
        # Split the section into subsections if necessary
        subsections = split_section_into_subsections(header, content, enc)

        # Combine adjacent tuples with less than 1000 tokens until they exceed 1000 tokens
        combined_subsections = combine_subsections(subsections)

        # Initialize the counter for numbering sequential identical subheaders
        subheader_count = 1

        # Process each combined subsection
        for subheader, subcontent in combined_subsections:
            # Update the subheader if there are multiple sequential identical subheaders
            if subheader_count > 1:
                subheader += f"-part{subheader_count}"
            subheader_count += 1

            # Use tiktoken to encode the subsection content as a sequence of tokens
            subcontent_tokens = enc.encode(subcontent)


            # Get the name of the output file
            #print("Subheader: ",subheader)
            section_name = subheader.replace(' ', '').replace('/','-')
            
            #print("Section name: ",section_name)
            output_path = f"{base_name}.{section_name}.txt"

            if (len(subcontent) == 0):
                subheader_count = subheader_count - 1            
            else:
                # Write the content to the output file
                with open(output_path, 'w') as f:
                    f.write(subcontent)
                print(f"{subheader} ({len(subcontent)} characters, {len(subcontent_tokens)} tokens) written to {output_path}")
                # Get the name of the summary file
                summary_path = f"{base_name}.{section_name}.summary.txt"
                # If the summary file does not exist, generate a summary
                if os.path.exists(summary_path):
                    print(f"Summary already exists at {summary_path}")
                else:
                    # Set the prompt for the summary
                    prompt = f"Please provide a detailed summary of the following paper section:\n{subcontent}\nPlease provide a detailed summary of the paper section above."
                    # Generate a summary for the subsection
                    summary = generate_summary(subcontent, prompt)
                    # Write the summary to a summary file
                    with open(summary_path, 'w') as f:
                        f.write(summary)
                    print(f"Summary written to {summary_path}")
        
        # If there is more than one summary file matching {base_name}.*{section_number}.summary.txt, generate a combined section summary
        section_number = section_name.split('.')[0]
        if len(glob.glob(f"{base_name}.{section_number}.*.summary.txt")) < 1:
            print(f"No summary files found for section {section_number}")
        elif len(glob.glob(f"{base_name}.{section_number}.*.summary.txt")) == 1:

            print(f"Only one summary file found for section {section_number}, promoting it to section summary")
            # Get the path of the summary file
            #print(glob.glob(f"{base_name}.{section_number}.*.summary.txt"))
            summary_path = glob.glob(f"{base_name}.{section_number}.*.summary.txt")[0]
            # Get the path of the section summary file
            section_summary_path = f"{base_name}.{section_name}.section_summary.txt"
            # Read the summary file and write it to the section summary file
            with open(summary_path, 'r') as f:
                summary = f.read()
            with open(section_summary_path, 'w') as f:
                f.write(summary)

            print(f"Summary promoted to section summary at {section_summary_path}")

        else:
            # Read in the section summaries
            summaries = []
            
            summary_pattern = f"{base_name}.*{section_number}.summary.txt"
            for summary_path in glob.glob(summary_pattern):
                with open(summary_path, 'r') as f:
                    summaries.append(f.read())
            # Concatenate the summaries into a single string
            subcontent = "\n\n".join(summaries)
            # Set the prompt for the overall section summary
            prompt = f"Please provide a detailed summary of the following sections:\n{subcontent}\nPlease provide a detailed summary of the sections above."
            # Get the path of the overall section summary file
            section_summary_path = f"{base_name}.{section_number}.section_summary.txt"
            # If the overall section summary file does not exist, generate a summary
            if os.path.exists(section_summary_path):
                print(f"Overall section summary already exists at {section_summary_path}")
            else:
                # Generate the overall section summary
                section_summary = generate_summary(content, prompt)
                # Write the overall section summary to a file
                with open(section_summary_path, 'w') as f:
                    f.write(section_summary)
                print(f"Overall section summary written to {section_summary_path}")



            
    # Check if the overall summary file already exists
    overall_summary_path = f"{base_name}.overall_summary.txt"
    if os.path.exists(overall_summary_path):
        print(f"Overall summary already exists at {overall_summary_path}")
    else:
        # Read in the abstract
        with open(f"{base_name}.Title-Abstract.txt", 'r') as f:
            abstract = f.read()
        # Read in the section summaries
        summaries = []
        summary_pattern = f"{base_name}.*.section_summary.txt"
        for summary_path in glob.glob(summary_pattern):
            with open(summary_path, 'r') as f:
                summaries.append(f.read())
        # Concatenate the abstract and summaries into a single string
        subcontent = abstract + "\n\n" + "\n\n".join(summaries)
        # Set the prompt for the overall summary
        prompt = f"Please provide a detailed summary of the following paper, based on its abstract and summaries of each section:\n{subcontent}\nPlease provide a detailed summary of the paper described above, based on the provided abstract and summaries of each section."
        # Generate the overall summary
        overall_summary = generate_summary(subcontent, prompt)
        # Write the overall summary to a file
        with open(overall_summary_path, 'w') as f:
            f.write(overall_summary)
        print(f"Overall summary written to {overall_summary_path}")
